comp/syntheticstestscheduler/impl/worker.go

```diff
@@ -0,0 +1,328 @@
+
+			if err = s.sendResult(&WorkerResult{
+				hostname:         hname,
+				tracerouteResult: result,
+				tracerouteError:  err,
+				testCfg:          syntheticsTestCtx,
+				triggeredAt:      triggeredAt,
+				startedAt:        startedAt,
+				finishedAt:       finishedAt,
+				duration:         duration,
+				tracerouteCfg:    tracerouteCfg,
+			}); err != nil {
+				s.log.Debugf("[worker%d] error sending result: %s", workerID, err)
+			}
+		}
+	}
+}
+
+// WorkerResult represents the result produced by a worker.
+type WorkerResult struct {
+	tracerouteResult payload.NetworkPath
+	tracerouteError  error
+	testCfg          SyntheticsTestCtx
+	triggeredAt      time.Time
+	startedAt        time.Time
+	finishedAt       time.Time
+	duration         time.Duration
+	tracerouteCfg    config.Config
+	hostname         string
+}
+
+// toNetpathConfig converts a SyntheticsTestConfig into a system-probe Config.
+func toNetpathConfig(c common.SyntheticsTestConfig) (config.Config, error) {
+	var cfg config.Config
+
+	switch payload.Protocol(c.Subtype) {
+	case payload.ProtocolUDP:
+		req, ok := c.Config.Request.(common.UDPConfigRequest)
+		if !ok {
+			return config.Config{}, fmt.Errorf("invalid UDP request type")
+		}
+		cfg.Protocol = payload.ProtocolUDP
+		cfg.DestHostname = req.Host
+		if req.Port != nil {
+			cfg.DestPort = uint16(*req.Port)
+		}
+		if req.MaxTTL != nil {
+			cfg.MaxTTL = uint8(*req.MaxTTL)
+		}
+		if req.Timeout != nil {
+			cfg.Timeout = time.Duration(*req.Timeout) * time.Second
+		}
+		if req.SourceService != nil {
+			cfg.SourceService = *req.SourceService
+		}
+		if req.DestinationService != nil {
+			cfg.DestinationService = *req.DestinationService
+		}
+
+	case payload.ProtocolTCP:
+		req, ok := c.Config.Request.(common.TCPConfigRequest)
+		if !ok {
+			return config.Config{}, fmt.Errorf("invalid TCP request type")
+		}
+		cfg.Protocol = payload.ProtocolTCP
+		cfg.DestHostname = req.Host
+		if req.Port != nil {
+			cfg.DestPort = uint16(*req.Port)
+		}
+		if req.MaxTTL != nil {
+			cfg.MaxTTL = uint8(*req.MaxTTL)
+		}
+		if req.Timeout != nil {
+			cfg.Timeout = time.Duration(*req.Timeout) * time.Second
+		}
+		cfg.TCPMethod = payload.TCPMethod(req.TCPMethod)
+		if req.SourceService != nil {
+			cfg.SourceService = *req.SourceService
+		}
+		if req.DestinationService != nil {
+			cfg.DestinationService = *req.DestinationService
+		}
+	case payload.ProtocolICMP:
+		req, ok := c.Config.Request.(common.ICMPConfigRequest)
+		if !ok {
+			return config.Config{}, fmt.Errorf("invalid ICMP request type")
+		}
+		cfg.Protocol = payload.ProtocolICMP
+		cfg.DestHostname = req.Host
+		if req.MaxTTL != nil {
+			cfg.MaxTTL = uint8(*req.MaxTTL)
+		}
+		if req.Timeout != nil {
+			cfg.Timeout = time.Duration(*req.Timeout) * time.Second
+		}
+		if req.SourceService != nil {
+			cfg.SourceService = *req.SourceService
+		}
+		if req.DestinationService != nil {
+			cfg.DestinationService = *req.DestinationService
+		}
```

Comment:
🤔 There is a lot of code duplication here, I think there is probably a way to factor that.